rllib/algorithms/algorithm.py

```diff
@@ -1588,46 +1596,69 @@ def add_policy(
                 returns False) will not be updated.
             evaluation_workers: Whether to add the new policy also
                 to the evaluation WorkerSet.
-            workers: A list of RolloutWorker/ActorHandles (remote
+            worker_list: A list of RolloutWorker/ActorHandles (remote
                 RolloutWorkers) to add this policy to. If defined, will only
                 add the given policy to these workers.
 
         Returns:
             The newly added policy (the copy that got added to the local
-            worker).
-        """
-
-        kwargs = dict(
-            policy_id=policy_id,
-            policy_cls=policy_cls,
-            observation_space=observation_space,
-            action_space=action_space,
-            config=config,
-            policy_state=policy_state,
-            policy_mapping_fn=policy_mapping_fn,
-            policies_to_train=list(policies_to_train) if policies_to_train else None,
-        )
-
-        def fn(worker: RolloutWorker):
-            # `foreach_worker` function: Adds the policy the the worker (and
-            # maybe changes its policy_mapping_fn - if provided here).
-            worker.add_policy(**kwargs)
+            worker). If `worker_list` was provided, None is returned.
 
+        Raises:
+            ValueError: If both `policy_cls` AND `policy` are provided.
+            KeyError: If the given `policy_id` already exists in this Algorithm.
+        """
+        # Deprecated args.
         if workers is not None:
-            ray_gets = []
-            for worker in workers:
-                if isinstance(worker, ActorHandle):
-                    ray_gets.append(worker.add_policy.remote(**kwargs))
-                else:
-                    fn(worker)
-            ray.get(ray_gets)
+            deprecation_warning(
+                old="Algorithm.add_policy(.., workers=...)",
+                new="Algorithm.add_policy(.., worker_list=...)",
+                error=False,
+            )
+            worker_list = workers
+
+        # Worker list is explicitly provided -> Use only those workers (local or remote)
+        # specified.
+        if worker_list is not None:
+            RolloutWorker.add_policy_to_workers(
```

Comment:
I like this change, this actually works well with my fault tolerance PR. I will make ``WorkerSet.add_policy_to_workers()`` the only way to go about this in the future actually.

one problem though, this should be ``WorkerSet.`` not ``RolloutWorker.``?

Reply:
Oh, yeah, great catch!